File: frontend/src/fileTypes.ts
/**
 * What an uploaded file is, and which pipeline it belongs to.
 *
 * Mirrors backend/specledger/file_types.py. The browser decides which
 * endpoint to call; the server decides whether to accept it. If the two
 * disagree, a file is either sent to the wrong pipeline or rejected after
 * upload with nothing useful said about why.
 *
 * Two kinds of upload exist and they are not interchangeable:
 *   catalogue — structured rows that become enriched 252-column records
 *   document  — a datasheet read for labelled specifications, creating no rows
 */

export type UploadKind = "catalogue" | "document" | "unsupported";

export interface UploadClassification {
  kind: UploadKind;
  extension: string;
  reason: string;
  /** What the format is, for the UI to explain itself. */
  description: string;
}

export const CATALOGUE_EXTENSIONS = [".csv", ".tsv", ".xlsx", ".json", ".xml"] as const;
export const DOCUMENT_EXTENSIONS = [".pdf", ".txt", ".docx", ".rtf"] as const;

const CATALOGUE_DESCRIPTIONS: Record<string, string> = {
  ".csv": "Comma-separated rows. The most common distributor feed.",
  ".tsv": "Tab-separated rows.",
  ".xlsx": "Excel workbook. The first sheet is read unless another is named.",
  ".json": "An array of product objects, or an object wrapping one.",
  ".xml": "Repeated elements, one per product.",
};

const DOCUMENT_DESCRIPTIONS: Record<string, string> = {
  ".pdf": "Manufacturer datasheets and specification sheets.",
  ".txt": "Plain text with no formatting.",
  ".docx": "Word document.",
  ".rtf": "Rich text.",
};

/** Refused, each with the reason and the remedy. */
const REJECTIONS: Record<string, string> = {
  ".xls": "Legacy Excel is not readable here — re-save it as .xlsx.",
  ".doc": "Legacy Word is not readable here — re-save it as .docx.",
  ".jpg": "An image carries no machine-readable text. Send the datasheet as a PDF.",
  ".jpeg": "An image carries no machine-readable text. Send the datasheet as a PDF.",
  ".png": "An image carries no machine-readable text. Send the datasheet as a PDF.",
  ".gif": "An image carries no machine-readable text. Send the datasheet as a PDF.",
  ".svg": "An image carries no machine-readable text. Send the datasheet as a PDF.",
  ".mp3": "Audio carries no product specifications SpecLedger can verify.",
  ".wav": "Audio carries no product specifications SpecLedger can verify.",
  ".mp4": "Video carries no product specifications SpecLedger can verify.",
  ".avi": "Video carries no product specifications SpecLedger can verify.",
  ".mov": "Video carries no product specifications SpecLedger can verify.",
  ".zip": "Archives are not opened. Upload the files inside it individually.",
  ".exe": "Executables are never accepted.",
  ".iso": "Disk images are never accepted.",
};

/** The file picker's filter — exactly what the app accepts. */
export const ACCEPT_ATTRIBUTE = [...CATALOGUE_EXTENSIONS, ...DOCUMENT_EXTENSIONS].join(",");

const SUPPORTED_LIST = [...CATALOGUE_EXTENSIONS, ...DOCUMENT_EXTENSIONS].join(", ");

function extensionOf(filename: string): string {
  const name = String(filename ?? "");
  const dot = name.lastIndexOf(".");
  // A dot in a directory name is not an extension.
  if (dot <= name.lastIndexOf("/") || dot === -1) return "";
  return name.slice(dot).toLowerCase();
}

export function classifyUpload(filename: string): UploadClassification {
  const extension = extensionOf(filename);

  if (extension in CATALOGUE_DESCRIPTIONS) {
    return { kind: "catalogue", extension, reason: "", description: CATALOGUE_DESCRIPTIONS[extension] };
  }
  if (extension in DOCUMENT_DESCRIPTIONS) {
    return { kind: "document", extension, reason: "", description: DOCUMENT_DESCRIPTIONS[extension] };
  }
  if (extension in REJECTIONS) {
    return { kind: "unsupported", extension, reason: REJECTIONS[extension], description: "" };
  }
  if (!extension) {
    return {
      kind: "unsupported", extension: "", description: "",
      reason: `The file has no extension, so its format is unknown. Supported: ${SUPPORTED_LIST}.`,
    };
  }
  return {
    kind: "unsupported", extension, description: "",
    reason: `'${extension}' is not a supported format. Supported: ${SUPPORTED_LIST}.`,
  };
}

File: backend/specledger/file_types.py
"""What an uploaded file is, and what SpecLedger will do with it.

Two kinds of upload reach this system and they are not interchangeable:

* a **catalogue** — structured rows that become enriched 252-column records
* a **document** — prose or a datasheet, read for labelled specifications
  and page-level evidence, producing no catalogue rows at all

Everything else is refused, and a refusal says why and what to send
instead. Silence here is expensive: a judge who uploads a scan of a
datasheet and sees nothing happen has no way to tell a rejected format
from a broken pipeline.

This module is the single place that rule lives. It had begun to spread
across ``ALLOWED_EXTENSIONS`` in the ingest endpoint, ``read_catalogue``'s
suffix branch, the intake endpoint's content-type check and three separate
pieces of UI copy — the same duplication that has produced the same defect
five times in this codebase.
"""

from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path


@dataclass(frozen=True)
class UploadKind:
    """What to do with an uploaded file."""
    kind: str            # "catalogue" | "document" | "unsupported"
    extension: str
    reason: str = ""     # why it was refused, and what to send instead
    description: str = ""  # what this format is, for the UI


# Structured data. Becomes rows.
CATALOGUE_FORMATS: dict[str, str] = {
    ".csv": "Comma-separated rows. The most common distributor feed.",
    ".tsv": "Tab-separated rows.",
    ".xlsx": "Excel workbook. The first sheet is read unless another is named.",
    ".json": "An array of objects, or an object wrapping one.",
    ".xml": "Repeated elements, one per product.",
}

# Prose and datasheets. Read for facts; creates no rows.
DOCUMENT_FORMATS: dict[str, str] = {
    ".pdf": "Manufacturer datasheets and specification sheets.",
    ".txt": "Plain text with no formatting.",
    ".docx": "Word document.",
    ".rtf": "Rich text.",
}

# Refused, with the reason the reader needs. A format that has a modern
# replacement names it rather than simply failing.
_REJECTIONS: dict[str, str] = {
    ".xls": "Legacy Excel is not readable here — re-save it as .xlsx.",
    ".doc": "Legacy Word is not readable here — re-save it as .docx.",
    ".jpg": "An image carries no machine-readable text. Send the datasheet as a PDF.",
    ".jpeg": "An image carries no machine-readable text. Send the datasheet as a PDF.",
    ".png": "An image carries no machine-readable text. Send the datasheet as a PDF.",
    ".gif": "An image carries no machine-readable text. Send the datasheet as a PDF.",
    ".svg": "An image carries no machine-readable text. Send the datasheet as a PDF.",
    ".mp3": "Audio carries no product specifications SpecLedger can verify.",
    ".wav": "Audio carries no product specifications SpecLedger can verify.",
    ".mp4": "Video carries no product specifications SpecLedger can verify.",
    ".avi": "Video carries no product specifications SpecLedger can verify.",
    ".mov": "Video carries no product specifications SpecLedger can verify.",
    ".zip": "Archives are not opened. Upload the files inside it individually.",
    ".exe": "Executables are never accepted.",
    ".iso": "Disk images are never accepted.",
}

_SUPPORTED_LIST = ", ".join(sorted(CATALOGUE_FORMATS) + sorted(DOCUMENT_FORMATS))


def classify_upload(filename: str) -> UploadKind:
    """Decide what an uploaded filename is, by extension.

    Extension only: sniffing content would let a mislabelled file take a
    path its uploader did not intend, and every real feed and datasheet
    arrives correctly named.
    """
    extension = Path(str(filename or "")).suffix.casefold()

    if extension in CATALOGUE_FORMATS:
        return UploadKind("catalogue", extension, description=CATALOGUE_FORMATS[extension])
    if extension in DOCUMENT_FORMATS:
        return UploadKind("document", extension, description=DOCUMENT_FORMATS[extension])
    if extension in _REJECTIONS:
        return UploadKind("unsupported", extension, reason=_REJECTIONS[extension])
    if not extension:
        return UploadKind(
            "unsupported", "",
            reason=f"The file has no extension, so its format is unknown. Supported: {_SUPPORTED_LIST}.",
        )
    return UploadKind(
        "unsupported", extension,
        reason=f"'{extension}' is not a supported format. Supported: {_SUPPORTED_LIST}.",
    )


def supported_extensions() -> tuple[str, ...]:
    """Every extension the system accepts, for the file picker's filter."""
    return tuple(sorted(CATALOGUE_FORMATS) + sorted(DOCUMENT_FORMATS))
